### ISSUE TITLE
Conflicting imports? Only in CLI, IDE is fine

### ISSUE THREAD

**Issue by quintesse** [ceylon/ceylon-compiler#2365]
This code

```ceylon
import org.w3c.dom.events {Event, EventListener}

class EventListenerExample() satisfies EventListener {
    shared actual void handleEvent(Event event) { 
    }
}
```

using this module descriptor:

```ceylon
native("jvm") module inexp.itsnatceylon "1.0.0" {
    import java.base "7";
    import javax.xml "7";
    shared import "xml-apis:xml-apis" "1.3.04";
}
```

results in these errors using the CLI compiler:

```
source/inexp/itsnatceylon/ceylonex.ceylon:2: error: imported declaration not found: 'Event'
import org.w3c.dom.events {Event, EventListener}
                           ^
source/inexp/itsnatceylon/ceylonex.ceylon:2: error: imported declaration not found: 'EventListener'
import org.w3c.dom.events {Event, EventListener}
                                  ^
source/inexp/itsnatceylon/ceylonex.ceylon:5: error: type declaration does not exist: 'Event'
    shared actual void handleEvent(Event event) {
                                   ^
source/inexp/itsnatceylon/ceylonex.ceylon:5: error: actual member does not refine any inherited member: 'handleEvent' in 'EventListenerExample'
    shared actual void handleEvent(Event event) {
                       ^
source/inexp/itsnatceylon/ceylonex.ceylon:4: error: type declaration does not exist: 'EventListener'
class EventListenerExample() satisfies EventListener {
```

but if you remove the `import javax.xml "7"` the errors disappear.

Even more strange is that in the IDE this compiles just fine (with or without the xml import)!

Found by @jmarranz

**Comment by quintesse** [comment]
Forgot to mention that both `java.xml` and `xml-apis:xml-apis` contain `org.w3c.dom.events {Event, EventListener}`

**Comment by quintesse** [comment]
If I debug what the typechecker does in the IDE then I can see that it *always* retrieves the `javax.xml` version of the classes *even* when the `java.xml` is removed and only the `xml-apis:xml-apis` is left in!

**Comment by quintesse** [comment]
Any ideas @FroMage ?

**Comment by FroMage** [comment]
Same package in multiple modules is just random at this point. Totally unsupported

**Comment by quintesse** [comment]
But does that mean we should at least try to give a coherent error? Or does "totally unsupported" mean "nothing we can do about it"?
And the fact that Eclipse seems just fine?
Or all part of this undefined behaviour?

**Comment by FroMage** [comment]
Right: random at this point.

**Comment by gavinking** [comment]
@FroMage so we're reassign this to 1.3, I assume?

Note that I also ran into the problem in #2373.

**Comment by jmarranz** [comment]
In my opinion a good Java integration is cornerstone for Ceylon use and
adoption, as you know there are no many examples of using Java libraries
inside Ceylon (regarding to leaving this problem to get fixed in 1.3).

I think you should provide a sort of optional class "excluding" pattern
when you import a Java jar to avoid double attempts of Java class loading
(this seems the current problem), if class duplication remains a problem.

Anyway in case you don't support .class duplication (not a big problem in
Java std because it just load the first one found), people need a correct
and more descriptive and precise error, the "not found" error is silly when
the real problem seems "I've found two or more .class files of Java X
class".

**Comment by gavinking** [comment]
> I think you should provide a sort of optional class "excluding" pattern when you import a Java jar to avoid double attempts of Java class loading

@jmarranz I *think* that is already possible via `overrides.xml`, right @FroMage?

**Comment by gavinking** [comment]
See here:

<http://ceylon-lang.org/documentation/1.1/reference/repository/maven/>

**Comment by jmarranz** [comment]
Right, I read it sometime ago but it seems complex and tightly related to
Maven, my first attempt of Ceylon was importing jars manually.

I mean something more simple like:

import "myjavalibdec" "ver"  { excluding : { org.w3c.dom.events.* }}

**Comment by jmarranz** [comment]
As far I remember I tried something using

http://ceylon-lang.org/documentation/reference/structure/module-properties/

But the problem has nothing to do with non required Maven dependencies.

To summarize the problem:

- Ceylon rigth now does not support the presence of duplicated .class
- java.xml seems to get imported ever or depending the context (bug?)
- java.xml does not include HTML DOM APIs, it forces to add another jar
containing all DOM stuff... missing a java.html?
- W3C DOM API is evolving, it seems legitimate to use APIS defined in
external jars in spite of classes already included in std Java. Because
this API is not core of Java, the extended API can be loaded instead of the
Java std provided (we are not trying to replace java.lang.String). This is
the reason of xml-apis jar in Maven exists.

**Comment by FroMage** [comment]
Yes, ATM you can exclude packages from a module with an overrides file, even for Ceylon modules, but I expect this to be more useful for Java modules. In the future we'll get to fixing package splits across multiple modules in a more satisfying way, but at this point there's not enough time for 1.2.

**Comment by jmarranz** [comment]
Of course I'm not going to decide your priorities but the most critical
problems are a non sense and confusing "not found" class error message when
the real error is the contrary (more than one class found), and missing
classes (ex HTML DOM API) in Java.X modularization.

 One case is a bug, and the other an incomplete feature. Both cases seem
easy to get fixed in release 1.2   :)

**Comment by jmarranz** [comment]
To be more accurated

http://ceylon-lang.org/documentation/1.1/reference/interoperability/java-from-ceylon/


   - javax.xml: contains javax.xml.**,org.w3c.dom.**, org.xml.sax.**


Seems not correct, org.w3c.dom.html.** are not included.

**Comment by gavinking** [comment]
@FroMage As @jmarranz says, there's definitely something wrong with the package-lists for `javax.xml` vs `oracle.jdk.jaxp`, but I can't figure out what the correct division should be.

**Comment by FroMage** [comment]
Right, the reference is not right, `oracle.jdk.jaxp` contains _some_ of `org.w3c.dom.*`. I will update it.

**Comment by jmarranz** [comment]
Hello again

Ceylon 1.2 release seems to fix this problem, but again some problems remain...

- javax.xml no longer is self-imported, fine! 

- javax.xml does not include org.w3c.dom.html, this is not problem, I can use a jar like xml-apis-1.3.04 and avoid importing javax.xml, but the documentation is wrong:

http://ceylon-lang.org/documentation/1.2/reference/interoperability/java-from-ceylon/

quote
javax.xml: contains javax.xml.**, org.w3c.dom, org.w3c.dom.{bootstrap,events,ls} (css, html, ranges, stylesheets, traversal and views are in the oracle.jdk.jaxp module), org.xml.sax.**
end quote

- if javax.xml AND xml-apis-1.3.04 are both imported AGAIN the error messages are like:

error: imported declaration not found: 'Event'

This message as said time before is *not correct*, because the true cause is *duplication" (same class in two "jars").

- I have another  problem, I have a jar with very few new classes under org.w3c.dom.* and org.w3c.dom.events.* these classes are NOT present in javax.xml (Java SDK) NOR in xml-apis-1.3.04.jar but when I include this jar I have again errors like:

error: imported declaration not found: 'Event'

It seems a conflict when detected there are TWO jars declaring classes into org.w3c.dom.events.* in spite of there is no class repetition in this case. When I remove the jar with the new classes the problem disappear.

In summary: 

1) The import system has problems or accepted restrictions (I can admit it) regarding to duplication of .class in different places AND regarding to the presence of the same package in several jars when importing different classes in the same package in spite of there is no duplication.

2) The message error is not valid, it does not explain the real problem in fact it confuses developers.

**Comment by FroMage** [comment]
Hi, could you add your comment to https://github.com/ceylon/ceylon/issues/2365 please? We've migrated all issues there.
